src/frontend/src/utils/stream_graph_formatter.rs

```diff
@@ -51,76 +61,77 @@ impl StreamGraphFormatter {
         }
     }
 
+    fn pretty_add_table<'a>(&mut self, tb: &Table) -> Pretty<'a> {
+        Pretty::debug(&self.add_table(tb))
+    }
+
     /// collect the table catalog and return the table id
     fn add_table(&mut self, tb: &Table) -> u32 {
         self.tables.insert(tb.id, tb.clone());
         tb.id
     }
 
-    fn explain_graph(
-        &mut self,
-        graph: &StreamFragmentGraph,
-        f: &mut impl std::fmt::Write,
-    ) -> std::fmt::Result {
+    fn explain_graph<'a>(&mut self, graph: &StreamFragmentGraph) -> Vec<Pretty<'a>> {
```

Comment:
The lifetime looks interesting 👀

Reply:
To me It should be sth like
```rust
fn explain_graph<'a>(&mut self, graph: &'a StreamFragmentGraph) -> Vec<Pretty<'a>> 
```
🤔 Why only adding it to the return type can work

Reply:
Because they're separate lifetime! That `'a` is only used in a Cow string